File: backend/app/services/query_interpreter.py
"""Query interpreter — classifies query intent and extracts entities."""

import re
import json
import logging
from dataclasses import dataclass

from app.providers.interface import GenerateTextOptions
from app.providers.router import get_provider_router

logger = logging.getLogger(__name__)

WHY_PAT = re.compile(r"\b(why|reason|rationale|purpose|explain|justify)\b", re.I)
WHEN_PAT = re.compile(r"\b(when|date|introduced|added|changed|removed|first|last)\b", re.I)
CHANGED_PAT = re.compile(r"\b(changed|modified|updated|diff|between|before|after)\b", re.I)
DEP_PAT = re.compile(r"\b(depend\w*|require\w*|import\w*|package|library|version|dependency)\b", re.I)
EDGE_PAT = re.compile(r"\b(edge.?case|special.?case|workaround|hack|legacy|compat|fallback)\b", re.I)
RAT_PAT = re.compile(r"\b(trade.?off|decision|chose|selected|picked|vs\.?|versus)\b", re.I)


@dataclass
class QueryInterpretation:
    intent: str
    entities: list[dict]
    time_hints: dict
    search_terms: list[str]
    confidence: float


def classify_intent_local(text: str) -> tuple[str, float]:
    scores = {
        "why": len(WHY_PAT.findall(text)),
        "when": len(WHEN_PAT.findall(text)),
        "what_changed": len(CHANGED_PAT.findall(text)),
        "dependency": len(DEP_PAT.findall(text)),
        "edge_case": len(EDGE_PAT.findall(text)),
        "rationale": len(RAT_PAT.findall(text)),
    }
    best = max(scores, key=scores.get)
    if scores[best] == 0:
        return "unknown", 0.3
    total = sum(scores.values())
    confidence = min(0.5 + (scores[best] / max(total, 1)) * 0.4, 0.9)
    return best, confidence


def extract_entities_local(text: str) -> list[dict]:
    entities = []
    for m in re.finditer(r"#(\d+)", text):
        entities.append({"type": "issue_or_pr", "value": m.group(1)})
    for m in re.finditer(r"\b([0-9a-f]{7,40})\b", text):
        entities.append({"type": "commit_hash", "value": m.group(1)})
    for m in re.finditer(r'"([^"]+)"', text):
        entities.append({"type": "symbol", "value": m.group(1)})
    for m in re.finditer(r"[\w/]+\.\w{1,5}", text):
        entities.append({"type": "file_path", "value": m.group(0)})
    for m in re.finditer(r"\b([A-Z][a-zA-Z]+(?:[A-Z][a-zA-Z]*)+)\b", text):
        entities.append({"type": "identifier", "value": m.group(1)})
    return entities


def extract_time_hints(text: str) -> dict:
    hints = {}
    for m in re.finditer(r"\b(\d{4})-(\d{2})-(\d{2})\b", text):
        hints["date"] = m.group(0)
    return hints


def extract_search_terms(text: str) -> list[str]:
    stop = {"why","does","this","the","is","was","has","have","had","do","did","it","that",
            "what","when","where","how","which","who","still","support","code","function",
            "method","class","file","module","a","an","in","on","for","of","with","by","from",
            "as","or","and","not","be","are","were","can","could","would","should"}
    return [w for w in re.findall(r"\b[a-zA-Z]\w+\b", text.lower()) if w not in stop and len(w) > 2]


class QueryInterpreter:
    def __init__(self):
        self.router = get_provider_router()

    async def interpret(self, query_text: str) -> QueryInterpretation:
        intent, confidence = classify_intent_local(query_text)
        entities = extract_entities_local(query_text)
        time_hints = extract_time_hints(query_text)
        search_terms = extract_search_terms(query_text)

        if confidence < 0.5:
            try:
                llm_result = await self._llm_classify(query_text)
                if llm_result:
                    return llm_result
            except Exception as e:
                logger.warning("LLM classification failed: %s", e)

        return QueryInterpretation(intent=intent, entities=entities, time_hints=time_hints,
                                   search_terms=search_terms, confidence=confidence)

    async def _llm_classify(self, query_text: str) -> QueryInterpretation | None:
        prompt = f'''Classify this developer query. Return JSON with:
- intent: one of "why", "when", "what_changed", "dependency", "edge_case", "rationale", "unknown"
- entities: list of type/value pairs
- search_terms: list of 2-5 key terms
Query: "{query_text}"
Return ONLY valid JSON.'''

        result = await self.router.generate_text(
            GenerateTextOptions(prompt=prompt, temperature=0.0, max_tokens=256),
            model_override="groq:llama-3.1-8b-instant",
        )
        try:
            data = json.loads(result.text.strip().removeprefix("```json").removesuffix("```").strip())
            return QueryInterpretation(
                intent=data.get("intent", "unknown"), entities=data.get("entities", []),
                time_hints={}, search_terms=data.get("search_terms", extract_search_terms(query_text)),
                confidence=0.7,
            )
        except (json.JSONDecodeError, KeyError):
            return None

File: src/lib/pipeline/query-interpreter.ts
import * as types from '@/types';
import { getRouter } from '@/lib/providers/router';

/**
 * QueryInterpreter classifies intent, resolves entities, and detects time hints.
 */
export class QueryInterpreter {
  async interpret(text: string): Promise<types.QueryInterpretation> {
    const router = getRouter();

    // 1. Rule-based entity extraction (fast, no LLM needed)
    const entities = this.extractEntities(text);
    const timeHints = this.extractTimeHints(text);

    // 2. LLM-based intent classification
    try {
      const result = await router.extractStructured({
        text,
        schema: {
          intent: 'why | when | what_changed | dependency | rationale | edge_case | unknown',
          searchTerms: ['string'],
          confidence: 0.0,
        },
        instructions: 'Classify the user query intent for a code archaeology tool. Extract key search terms. Return JSON.',
        model: router.getModelForStage('queryClassification'),
      });

      const intent = (result.data.intent as types.QueryIntent) || 'unknown';
      const searchTerms = (result.data.searchTerms as string[]) || text.split(/\s+/).filter(t => t.length > 2);
      const confidence = (result.data.confidence as number) || 0.5;

      return { intent, entities, timeHints, searchTerms, confidence };
    } catch {
      // Fallback to rule-based classification
      const intent = this.classifyIntent(text);
      // Use all meaningful words (>= 2 chars) as search terms for better recall
      const stopWords = new Set(['the', 'a', 'an', 'is', 'are', 'was', 'were', 'be', 'been', 'being',
        'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should', 'may',
        'might', 'shall', 'can', 'to', 'of', 'in', 'for', 'on', 'with', 'at', 'by', 'from',
        'as', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'between',
        'out', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'this', 'that',
        'these', 'those', 'and', 'but', 'or', 'nor', 'not', 'so', 'what', 'which', 'who',
        'when', 'where', 'how', 'all', 'each', 'every', 'both', 'few', 'more', 'most',
        'other', 'some', 'such', 'than', 'too', 'very', 'just', 'about', 'also', 'only']);
      const searchTerms = text
        .toLowerCase()
        .replace(/[^a-z0-9\s._/-]/g, ' ')
        .split(/\s+/)
        .filter(t => t.length >= 2 && !stopWords.has(t));
      return { intent, entities, timeHints, searchTerms, confidence: 0.4 };
    }
  }

  private extractEntities(text: string): Array<{ type: string; value: string }> {
    const entities: Array<{ type: string; value: string }> = [];

    // Issue/PR numbers
    for (const m of text.matchAll(/#(\d+)/g)) {
      entities.push({ type: 'issue_number', value: m[1] });
    }

    // Commit hashes
    for (const m of text.matchAll(/\b([a-f0-9]{7,40})\b/g)) {
      if (!/^\d{4}-\d{2}-\d{2}$/.test(m[1])) {
        entities.push({ type: 'commit_hash', value: m[1] });
      }
    }

    // Quoted symbols
    for (const m of text.matchAll(/['"](\w+)['"]/g)) {
      entities.push({ type: 'symbol', value: m[1] });
    }

    // PascalCase (symbols/classes)
    for (const m of text.matchAll(/\b([A-Z][a-z]+[A-Z]\w+)\b/g)) {
      if (!entities.some(e => e.value === m[1])) {
        entities.push({ type: 'symbol', value: m[1] });
      }
    }

    // File paths
    for (const m of text.matchAll(/\b([\w./\\-]+\.[a-z]{2,4})\b/g)) {
      if (m[1].includes('/') || m[1].includes('\\')) {
        entities.push({ type: 'file_path', value: m[1] });
      }
    }

    return entities;
  }

  private extractTimeHints(text: string): Record<string, unknown> {
    const hints: Record<string, unknown> = {};

    // Version references
    const versionMatch = text.match(/v?(\d+\.\d+(?:\.\d+)?)/);
    if (versionMatch) hints.version = versionMatch[1];

    // Year references
    const yearMatch = text.match(/\b(20\d{2})\b/);
    if (yearMatch) hints.year = parseInt(yearMatch[1]);

    // Relative time references
    if (/\b(recent|latest|new)\b/i.test(text)) hints.recency = 'recent';
    if (/\b(old|legacy|original)\b/i.test(text)) hints.recency = 'old';
    if (/\b(future|upcoming|next)\b/i.test(text)) hints.recency = 'future';

    return hints;
  }

  private classifyIntent(text: string): types.QueryIntent {
    const lower = text.toLowerCase();

    if (/^why\b/.test(lower) || /\bwhy\b/.test(lower)) return 'why';
    if (/^when\b/.test(lower) || /\bwhen\b/.test(lower)) return 'when';
    if (/\bwhat changed\b/.test(lower) || /\bwhat.*change/.test(lower)) return 'what_changed';
    if (/\bdepend(ency|s|ent)\b/.test(lower)) return 'dependency';
    if (/\btradeoff|rationale|why|reason|motivation\b/.test(lower)) return 'rationale';
    if (/\bedge\s*case|exception|error case\b/.test(lower)) return 'edge_case';

    return 'unknown';
  }

  /** Check if the query is asking for a diagram/architecture visualization */
  static isDiagramQuery(text: string): boolean {
    const lower = text.toLowerCase();
    return /\b(diagram|flowchart|architecture|arch\b|system design|data flow|component|module|class diagram|sequence diagram|er diagram|erd|mermaid|visuali[sz]e|draw|graph|tree|hierarchy|dependency graph|call graph|overview|structure)\b/.test(lower);
  }
}
